### DISCUSSION TITLE
Best way to insert / update relations

### DISCUSSION THREAD

**Discussion by Tristannos** [serverpod/serverpod#1717] [Q&A] [answered]
I just updated my serverpod to 1.2 and started out with the relations (with Object)

I have these 2 tables, what is the best way to update the location object from the Additional_user_info endpoint.
The only way I see it, is this the way to connect those 2 and update them together with the Additional_user_info table:

```
if(additionalUserInfo.location != null && additionalUserInfo.location!.id == null && additionalUserInfo.locationId == null){
         // Inserting the new row, with a connection to the user
        Location location = await Location.db.insertRow(session, additionalUserInfo.location!);
        additionalUserInfo.locationId = location.id;
    }else{
      // Updating the row
      await Location.db.updateRow(session, additionalUserInfo.location!);
    }
```

I don't think this is clean code if I have to do this with every relation I have.
instead isn't it way easier and cleaner to have it like this and update the `Location`  relation with it?

`await AdditionalUserInfo.db.updateRow(session, additionalUserInfo);
`

Or when I attach it, it stil does not update:

`await AdditionalUserInfo.db.attachRow.location(session, additionalUserInfo, additionalUserInfo.location!);
`

**Additional_user_info**
<img width="351" alt="SCR-20240106-oqhx" src="https://github.com/serverpod/serverpod/assets/55278228/40334995-5e0d-4661-b616-5fe0b0c8f07d">

**Location**
<img width="199" alt="SCR-20240106-oqll" src="https://github.com/serverpod/serverpod/assets/55278228/a964479e-b921-40e5-a69e-4231ad5b6b21">

**Comment by Isakdl** [comment, marked as answer]
Hey @Tristannos! 

Yes it would be nice to be able to insert / update deeply nested objects. This feature is not implemented yet but is tracked here: https://github.com/serverpod/serverpod/issues/1343. 

Your first example has an issue where the else case could crash if location is null, but can easily be fixed like so:
```dart
if(additionalUserInfo.location != null) {
  if(additionalUserInfo.location!.id == null && additionalUserInfo.locationId == null) {
       // Inserting the new row, with a connection to the user
      Location location = await Location.db.insertRow(session, additionalUserInfo.location!);
      additionalUserInfo.locationId = location.id;
  } else {
    // Updating the row
    await Location.db.updateRow(session, additionalUserInfo.location!);
  }
}
```

I think the above problem is solved with an upsert feature, also yet to be implemented but tracked here: https://github.com/serverpod/serverpod/issues/232.

Right now you have to handle each object as a separate entity when storing them to the db.
